app/services/bulk_update/recommendations_uploads/validate_trainee.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module BulkUpdate
+  module RecommendationsUploads
+    # finds and validates a trainee scoped within provider, via trn/hesa/provider_trainee_id
+    class ValidateTrainee
+      def initialize(row:, provider:)
+        @row = row
+        @provider = provider
+        @messages = []
+
+        find!
+        validate!
+      end
+
+      def trainee
+        @trainees.first if @trainees.size == 1
+      end
+
+      def valid?
+        @messages.empty?
+      end
+
+      attr_reader :messages
+
+    private
+
+      attr_reader :row, :provider
+
+      def find!
+        trainees_by_trn!
+        trainees_by_hesa_id!
+        trainees_by_provider_trainee_id!
+      end
+
+      def validate!
+        @messages << "Multiple trainees found via #{found_with}" if @trainees.size > 1
+        @messages << "No trainee could be matched" if @trainees.empty?
+      end
+
+      def found_with
+        { trn: "TRN", hesa_id: "HESA ID", provider_trainee_id: "the provider trainee ID" }[@found_with]
+      end
+
+      # rubocop:disable Naming/MemoizedInstanceVariableName
+      def trainees_by_trn!
+        return unless row.trn =~ /^\d{7}$/
+
+        @trainees ||= begin
+          t = provider.trainees.where(trn: row.trn)
```

Comment:
For this and equivalent lines in `trainees_by_hesa_id!` and `trainees_by_provider_trainee_id!`

The error spreadsheet says: "multiple trainee records found for provider based on TRN, or HESA ID in an active state"

Does this mean we need to filter by `trn_received` here?

Reply:
if I do that then the find my hesa or provider trainee (should) never run because any trainees that have those but don't have a trn woudl be filtered out ¯\_(ツ)_/¯ another thing I'm hoping to clear up with john oates

Reply:
Some of these might be better to deal with with a chat, agree!